### DISCUSSION TITLE
http connection with body credentials

### DISCUSSION THREAD

**Discussion by alexisaraya** [apache/airflow#26268] [General] [answered]
Hey Guys,
Its posible storing http conection in secret value with extras body credentias?
I using http conection but in login send user crendentias in body.

`hook = HttpHook(http_conn_id="autentication_api", method="POST",)
        logging.info("Calling HTTP method")
        response = hook.run(endpoint='/user/login',data=json.dumps({"data":{"email":"api_login@xxxxxx", "password": "xxxxxx"}}),
        headers={"Content-Type":"application/json"},)`

I currently send the credentials in the code, but I would like them to be in the key manager next to the connection.

Tranks for your help.

**Comment by potiuk** [comment, marked as answer]
Yes  - you need your server to support basic authentication (default) or any other AuthType supported by requests library
. If you need other auth than those supported by requests library, you need to write your own custom Hook extending HTTPHook and implementing any authentication scheme - same as you would in any other Python code using requests library.

You can see how auth is instantiated by the Hook https://github.com/apache/airflow/blob/main/airflow/providers/http/hooks/http.py#L97 

And you can read all details (including examples) on how you can write your own custom authentication if the authentication schemes provided by requests are not enough for you  https://requests.readthedocs.io/en/latest/user/advanced/#custom-authentication

**Comment by jose-sotelocohen-adyen** [comment]
@potiuk I am doing this now, but I do need to add an extra field for the HTTP connection type so I can add secret token BS3PasswordFieldWidget.

I've tried adding get_connection_form_widgets to my subclass of HttpHook but it doesn't work, do you have any clues on how to make this work?

**Reply by potiuk** [reply]
No. I do not know how it "does not work" so I am not sure how I can help. And anyhow - just looking at other code should guide you in this.  You need to experiment with it a bit - change things here and there and see if it helps,. Don't forget to restart things when you change anything.

**Reply by jose-sotelocohen-adyen** [reply]
Sounds good! thanks :)
